### PR TITLE
feat(stdlib): add punycode encoding functions

### PR DESCRIPTION
**Pull request by esensar**
This adds `encode_punycode` and `decode_punycode` functions. It also adds tests to confirm `parse_url` function behavior when it comes to punycode.

Fixes: #659

### CODE REVIEW

**Comment by esensar** [comment]
This adds punycode related functions. I have also realised that `parse_url` already handles that, so I have added a few tests to confirm it. The newly added crate for handling punycode was already a transitive dependency of the project (because of the use of `url` crate, and `idna` crate which handles punycode is a direct dependency of `url` and maintained in the same repository https://github.com/servo/rust-url).

**Comment by johnhtodd** [comment]
I haven't tested this patch, so forgive me if this lazy request could be answered that way: What happens if a string not containing UTF-8 characters is run through the encode_punycode process?  Or a non-punycode string is run through decode_punycode?

**Comment by esensar** [comment]
Right, I should have added some tests and examples to demonstrate that behavior. To answer your question, it will only encode whatever requires encoding, so for that reason the example (www.café.com) only encodes the café part.

I will add tests and examples with fully ASCII domains to make this more clear.

**Comment by johnhtodd** [comment]
Thanks - more test cases are always good.  Very happy to see this feature.  I suspect it will have a better chance of approval as a PR if it also had documentation.

**Review by pront — COMMENTED** [review]
Thank you @esensar, this looks good. 

Can you also add a VRL test [like this](https://github.com/vectordotdev/vrl/blob/main/lib/tests/tests/functions/parse_groks.vrl) that exercises both encode and decode?

**Comment by jszwedko** [comment]
> Thank you @esensar, this looks good.
> 
> Can you also add a VRL test [like this](https://github.com/vectordotdev/vrl/blob/main/lib/tests/tests/functions/parse_groks.vrl) that exercises both encode and decode?

Additionally, could we add these new functions to the benchmarks in `benches/stdlib.rs`?

**Comment by esensar** [comment]
> > Thank you @esensar, this looks good.
> > Can you also add a VRL test [like this](https://github.com/vectordotdev/vrl/blob/main/lib/tests/tests/functions/parse_groks.vrl) that exercises both encode and decode?
> 
> Additionally, could we add these new functions to the benchmarks in `benches/stdlib.rs`?

Yeah, I have added them now.

I have added both cases when there is nothing to encode/decode and with IDN strings, to confirm that cases when no encoding is needed are faster (an order of magnitude faster on my machine).

**Comment by esensar** [comment]
I have added VRL tests as well, demonstrating encoding, decoding and combination with `parse_url` as well. I realized that I have made the function infallible accidentally, so I have updated that too.

File: lib/tests/tests/functions/punycode/encode_decode.vrl [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# object: { "host": "www.ليهمابتكلموشعربي؟.他们为什么不说中文" }
+# result: "www.ليهمابتكلموشعربي؟.他们为什么不说中文"
+
+encoded = encode_punycode!(.host)
+decode_punycode!(encoded)
```
Comment (pront) [inline lib/tests/tests/functions/punycode/encode_decode.vrl:5]: Thanks for adding this, looks good.

File: src/stdlib/decode_punycode.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    value: Box<dyn Expression>,
+}
+
+impl FunctionExpression for DecodePunycodeFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let string = value.try_bytes_utf8_lossy()?;
+
+        let (encoded, result) = idna::domain_to_unicode(&string);
+        result.map_err(|err| format!("unable to decode punycode: {err}"))?;
```
Comment (pront) [inline src/stdlib/decode_punycode.rs:57]: 🥜 rename `err` to `errors`

File: src/stdlib/encode_punycode.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    value: Box<dyn Expression>,
+}
+
+impl FunctionExpression for EncodePunycodeFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let string = value.try_bytes_utf8_lossy()?;
+
+        let encoded = idna::domain_to_ascii(&string)
+            .map_err(|err| format!("unable to encode to punycode: {err}"))?;
```
Comment (pront) [inline src/stdlib/encode_punycode.rs:62]: 🥜 rename `err` to `errors`

**Review by pront — APPROVED** [review]
Thank you for adding the tests!

**Pull request merged**